MetaMask/metamask-extension
# Feat/collectibles send flow

This PR enables sending of ERC721 tokens from within the final piece of bringing the extension to parity with mobile for NFT support. ERC1155 sends are not yet enabled (as is the case on mobile).
[Relevant Figma Designs](https://www.figma.com/file/8Xe22jEPgcElS5tWPw8AGM/NFTs-in-extension-(MVP)?node-id=1%3A3)

Screenshots:
Stepping through the Send Flow:
<img width="662" alt="Screen Shot 2022-01-04 at 4 47 39 PM" src="https://user-images.githubusercontent.com/34557516/148134887-993edb08-3824-4d46-805d-521aa42fef79.png">
<img width="503" alt="Screen Shot 2022-01-04 at 4 51 08 PM" src="https://user-images.githubusercontent.com/34557516/148134896-324bfb93-7bc7-4d7b-ac44-f41577a97971.png">
<img width="493" alt="Screen Shot 2022-01-04 at 4 51 18 PM" src="https://user-images.githubusercontent.com/34557516/148134900-1331b9eb-e4f4-42fa-a91f-1116f1431cea.png">
<img width="952" alt="Screen Shot 2022-01-04 at 4 51 31 PM" src="https://user-images.githubusercontent.com/34557516/148134906-80e6549e-3eb3-46fe-af44-51ef69ae09d1.png">

If the user does not own the collectible (if it is in the "Previously Owned" Section) they can still click into view the details,
but they will not see a send button available:
<img width="958" alt="Screen Shot 2022-01-04 at 5 00 33 PM" src="https://user-images.githubusercontent.com/34557516/148135603-77f39e65-2b6e-4a60-b1cb-6d243f8f5f70.png">
<img width="952" alt="Screen Shot 2022-01-04 at 4 51 54 PM" src="https://user-images.githubusercontent.com/34557516/148134999-729ad5aa-46fa-45ba-a4cf-792ef0b2f3e6.png">

And if the token is an ERC1155 we show but disable the send button with an info tooltip to indicate that we don't yet support sending this type of token:
<img width="1002" alt="Screen Shot 2022-01-04 at 4 52 19 PM" src="https://user-images.githubusercontent.com/34557516/148135084-fa20da03-722d-469d-8e03-641fd8c38059.png">

Manual testing steps:  
  - Add an NFT to your wallet on a test net (you can use this [faucet](https://faucet.paradigm.xyz/))
  - Click into the details page of the NFT and click the send button
  - Go through the send flow as you would for a token.
  - Verify that the NFT has sent to the address you intended.
  - Click back into the token details page, the send button should disappear, and when you return back up to the collectibles tab, that NFT should be removed from your primary collection and added to the "Previously Owned" section.
  
 **Related (and non-blocking in my opinion) work still left to do:**
 - Migrate or add notification tell users they will have to manually migrate their ERC721/1155 tokens via the `Import NFT` flow.
 - Add validation prevent users from adding ERC721/1155 tokens through the add tokens flow (and direct them to the `Import NFT` flow) 
 - Create an NFT flow in the test-dapp
 - [This related but not blocking bug](https://github.com/MetaMask/metamask-extension/issues/11611) for which I have PRs [here](https://github.com/MetaMask/controllers/pull/667) and [here (WIP)](https://github.com/MetaMask/metamask-extension/pull/13131) 
 - ERC1155 sending
